src/database/connection.ts

```diff
@@ -0,0 +1,55 @@
+import 'reflect-metadata';
+import { createConnection, ConnectionOptions } from 'typeorm';
+import { NotablePerson } from './entities/notablePerson';
+import { NotablePersonEvent } from './entities/event';
+import { NotablePersonLabel } from './entities/label';
+import { NotablePersonEventComment } from './entities/comment';
+import { User } from './entities/user';
+import { readJson } from '../helpers/readFile';
+import { isUsingProductionDatabase } from './env';
+
+const {
+  // These variables are for the development database
+  // provided by AWS Elastic Beanstalk
+  RDS_DB_NAME,
+  RDS_HOSTNAME,
+  RDS_PORT,
+  RDS_USERNAME,
+  RDS_PASSWORD,
+} = process.env;
+
+const getConfig = async (): Promise<ConnectionOptions> => ({
+  type: 'mysql',
+
+  ...isUsingProductionDatabase
+    ? {
+        ...await readJson<DatabaseConfig>('secrets/db.production.json'),
+
+        // Do not set any of these to `true`,
+        // otherwise the database will be destroyed
+        synchronize: false,
+        dropSchema: false,
+      }
+    : {
+        database: RDS_DB_NAME || 'hollowverse',
+        host: RDS_HOSTNAME || 'localhost',
+        port: Number(RDS_PORT) || 3306,
+        username: RDS_USERNAME || 'root',
+        password: RDS_PASSWORD || '123456',
+        synchronize: true,
+        dropSchema: false,
+      },
+});
+
+export const connection = getConfig().then(async config =>
```

Comment:
I used to think that it's not actually needed. But it's now enforced by a TSLint rule. The rationale makes sense to me. Suppose you have a function that returns a promise, but also does something synchronus:

```javascript
function validate() {
   // Do something synchronous
   return validateAsync(this);
}
```

Somewhere else we use this promise with `.then` and `.catch`:

```javascript
validate.then(/* ... */).catch(e => console.error(e));
```

If the synchronous operation in `validate()` throws an error, it won't be handled by the `catch` callback above. We would need to do:

```
try {
  validate.then(/* ... */).catch(e => console.error(e));
} catch (e) {
   console.error(e)
}
```

So we need to handle errors in two different ways.

By marking the function as `async`, it effectively becomes:

```
function validate() {
  return Promise.resolve(() => {
    // Do something synchronous
   return validateAsync(this);
  });
}
```

Now all errors are converted to rejections, whether they are thrown synchronously or asynchronously, and the `catch` handler catches both errors.

The TSLint rule enforces that any function that returns a promise must be marked `async` so that all errors are handled in a unified way.

Reply:
Interesting